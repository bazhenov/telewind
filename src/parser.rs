use std::fmt::Display;

use chrono::{DateTime, FixedOffset, TimeZone};
use chrono_tz::Asia::Vladivostok;
use lazy_static::lazy_static;
use regex::Regex;
use select::{
    document::Document,
    node::Find,
    predicate::{Name, Predicate},
};
use serde::{Deserialize, Serialize};

lazy_static! {
    static ref WIND_DIRECTION: Regex = Regex::new("([0-9]{1,3})°").unwrap();
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Observation {
    pub time: DateTime<FixedOffset>,
    pub direction: u16,
    pub avg_speed: f32,
}

impl Display for Observation {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{} {:2.1} m/s {:3}°",
            self.time.format("%H:%M"),
            self.avg_speed,
            self.direction
        )
    }
}

pub fn parse(input: &str) -> Vec<Observation> {
    let document = Document::from(input);
    let selector = Name("table").descendant(Name("tr"));
    let rows = document.find(selector);
    let mut result = vec![];

    for row in rows {
        if let Some(_) = row.find(Name("th")).next() {
            // skippping header
            continue;
        }
        let mut columns = row.find(Name("td"));
        let time = parse_column(&mut columns, vlat_time_parser);
        let direction = parse_column(&mut columns, direction_parser);
        let avg_speed = parse_column(&mut columns, wind_speed_parser);

        match (time, direction, avg_speed) {
            (Some(time), Some(direction), Some(avg_speed)) => result.push(Observation {
                time,
                direction,
                avg_speed,
            }),
            _ => {
                panic!("Unable to parse: {}", row.html())
            }
        }
    }
    result
}

fn parse_column<O, I: Predicate>(
    columns: &mut Find<I>,
    parser: fn(&str) -> Option<O>,
) -> Option<O> {
    columns.next().and_then(|n| parser(&n.text()))
}

// Parsing the string of format: `СЗЗ (301°)`
fn direction_parser(input: &str) -> Option<u16> {
    if let Some(caps) = WIND_DIRECTION.captures(input) {
        let direction = caps.get(1).unwrap();
        let direction = direction.as_str().parse::<u16>().unwrap();
        if direction <= 360 {
            return Some(direction % 360);
        }
    }
    None
}

fn wind_speed_parser(input: &str) -> Option<f32> {
    input.parse::<f32>().ok()
}

// Parse time in format: `29.10.2022 22:45` assuming it's in VLAT
fn vlat_time_parser(input: &str) -> Option<DateTime<FixedOffset>> {
    Vladivostok
        .datetime_from_str(input, "%d.%m.%Y %H:%M")
        .ok()
        .map(|t| t.with_timezone(&FixedOffset::east(10 * 3600)))
}

#[cfg(test)]
mod test {

    use super::*;
    use insta::assert_yaml_snapshot;

    #[test]
    fn foo() {
        let input = include_str!("../tests/example.html");
        assert_yaml_snapshot!(parse(input));
    }
}
