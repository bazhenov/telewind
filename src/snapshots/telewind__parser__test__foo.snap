---
source: src/parser.rs
expression: parse(input)
---
- time: "2022-10-29T22:46:00+10:00"
  direction: 318
  avg_speed: 2.6
  max_speed: 3.7
  min_speed: 1.8
- time: "2022-10-29T22:45:00+10:00"
  direction: 301
  avg_speed: 2.4
  max_speed: 3
  min_speed: 1.5
- time: "2022-10-29T22:44:00+10:00"
  direction: 314
  avg_speed: 2.8
  max_speed: 3.7
  min_speed: 1.4
- time: "2022-10-29T22:43:00+10:00"
  direction: 313
  avg_speed: 3
  max_speed: 3.5
  min_speed: 2.6
- time: "2022-10-29T22:42:00+10:00"
  direction: 311
  avg_speed: 2.7
  max_speed: 3.4
  min_speed: 2.1
- time: "2022-10-29T22:41:00+10:00"
  direction: 322
  avg_speed: 2.5
  max_speed: 3.5
  min_speed: 1.6
- time: "2022-10-29T22:40:00+10:00"
  direction: 289
  avg_speed: 2.4
  max_speed: 3.4
  min_speed: 1.6
- time: "2022-10-29T22:38:00+10:00"
  direction: 310
  avg_speed: 2.3
  max_speed: 2.6
  min_speed: 1.8
- time: "2022-10-29T22:37:00+10:00"
  direction: 313
  avg_speed: 2.6
  max_speed: 3.4
  min_speed: 1.7
- time: "2022-10-29T22:36:00+10:00"
  direction: 308
  avg_speed: 2.2
  max_speed: 2.8
  min_speed: 1.3
- time: "2022-10-29T22:35:00+10:00"
  direction: 324
  avg_speed: 2.3
  max_speed: 3.3
  min_speed: 1.5
- time: "2022-10-29T22:34:00+10:00"
  direction: 313
  avg_speed: 2.6
  max_speed: 3.1
  min_speed: 2.1
- time: "2022-10-29T22:33:00+10:00"
  direction: 318
  avg_speed: 2.5
  max_speed: 3
  min_speed: 1.8
- time: "2022-10-29T22:32:00+10:00"
  direction: 317
  avg_speed: 2
  max_speed: 3
  min_speed: 1
- time: "2022-10-29T22:31:00+10:00"
  direction: 312
  avg_speed: 2.1
  max_speed: 2.7
  min_speed: 1.3
- time: "2022-10-29T22:30:00+10:00"
  direction: 316
  avg_speed: 2.1
  max_speed: 2.8
  min_speed: 1.1
- time: "2022-10-29T22:29:00+10:00"
  direction: 316
  avg_speed: 3
  max_speed: 3.6
  min_speed: 2.3
- time: "2022-10-29T22:28:00+10:00"
  direction: 314
  avg_speed: 2.7
  max_speed: 3.4
  min_speed: 1.8
- time: "2022-10-29T22:27:00+10:00"
  direction: 308
  avg_speed: 2.9
  max_speed: 2.9
  min_speed: 2.3
- time: "2022-10-29T22:26:00+10:00"
  direction: 313
  avg_speed: 2.9
  max_speed: 3.6
  min_speed: 2.4
- time: "2022-10-29T22:25:00+10:00"
  direction: 305
  avg_speed: 2.5
  max_speed: 3.6
  min_speed: 1.9
- time: "2022-10-29T22:23:00+10:00"
  direction: 323
  avg_speed: 2.7
  max_speed: 3.6
  min_speed: 1.9
- time: "2022-10-29T22:22:00+10:00"
  direction: 318
  avg_speed: 2.6
  max_speed: 3.2
  min_speed: 2
- time: "2022-10-29T22:21:00+10:00"
  direction: 311
  avg_speed: 3.3
  max_speed: 4.5
  min_speed: 2.6
- time: "2022-10-29T22:20:00+10:00"
  direction: 320
  avg_speed: 2.9
  max_speed: 3.9
  min_speed: 2.2
- time: "2022-10-29T22:19:00+10:00"
  direction: 330
  avg_speed: 2.8
  max_speed: 3.7
  min_speed: 1.7
- time: "2022-10-29T22:18:00+10:00"
  direction: 318
  avg_speed: 2.2
  max_speed: 3.3
  min_speed: 1
- time: "2022-10-29T22:17:00+10:00"
  direction: 318
  avg_speed: 3.1
  max_speed: 4.5
  min_speed: 1.7
- time: "2022-10-29T22:16:00+10:00"
  direction: 318
  avg_speed: 3.2
  max_speed: 4.2
  min_speed: 1.7
- time: "2022-10-29T22:15:00+10:00"
  direction: 318
  avg_speed: 2.2
  max_speed: 3
  min_speed: 1.2
- time: "2022-10-29T22:14:00+10:00"
  direction: 306
  avg_speed: 3.2
  max_speed: 4.1
  min_speed: 1.9
- time: "2022-10-29T22:13:00+10:00"
  direction: 316
  avg_speed: 3
  max_speed: 4.4
  min_speed: 1.7
- time: "2022-10-29T22:12:00+10:00"
  direction: 344
  avg_speed: 3.1
  max_speed: 4.5
  min_speed: 1.7
- time: "2022-10-29T22:11:00+10:00"
  direction: 317
  avg_speed: 3.6
  max_speed: 5
  min_speed: 1.9
- time: "2022-10-29T22:10:00+10:00"
  direction: 299
  avg_speed: 2.8
  max_speed: 6.5
  min_speed: 1.5
- time: "2022-10-29T22:09:00+10:00"
  direction: 305
  avg_speed: 3.1
  max_speed: 4.5
  min_speed: 2
- time: "2022-10-29T22:07:00+10:00"
  direction: 309
  avg_speed: 3.4
  max_speed: 4.6
  min_speed: 2.5
- time: "2022-10-29T22:06:00+10:00"
  direction: 317
  avg_speed: 4.4
  max_speed: 5.1
  min_speed: 3.2
- time: "2022-10-29T22:05:00+10:00"
  direction: 285
  avg_speed: 3.5
  max_speed: 4.1
  min_speed: 2.3
- time: "2022-10-29T22:04:00+10:00"
  direction: 304
  avg_speed: 3.7
  max_speed: 4.7
  min_speed: 2.7
- time: "2022-10-29T22:03:00+10:00"
  direction: 309
  avg_speed: 3.9
  max_speed: 5.2
  min_speed: 2.9
- time: "2022-10-29T22:02:00+10:00"
  direction: 298
  avg_speed: 3.9
  max_speed: 5
  min_speed: 2.4
- time: "2022-10-29T22:01:00+10:00"
  direction: 316
  avg_speed: 3
  max_speed: 4.3
  min_speed: 1.9
- time: "2022-10-29T22:00:00+10:00"
  direction: 321
  avg_speed: 3.4
  max_speed: 4.2
  min_speed: 2.3
- time: "2022-10-29T21:59:00+10:00"
  direction: 297
  avg_speed: 3.3
  max_speed: 5
  min_speed: 1.3
- time: "2022-10-29T21:58:00+10:00"
  direction: 312
  avg_speed: 4
  max_speed: 5.5
  min_speed: 2.8
- time: "2022-10-29T21:57:00+10:00"
  direction: 309
  avg_speed: 4.1
  max_speed: 6.1
  min_speed: 1.9
- time: "2022-10-29T21:56:00+10:00"
  direction: 321
  avg_speed: 4.2
  max_speed: 5.2
  min_speed: 3.3
- time: "2022-10-29T21:55:00+10:00"
  direction: 302
  avg_speed: 4.1
  max_speed: 5.2
  min_speed: 3
- time: "2022-10-29T21:54:00+10:00"
  direction: 303
  avg_speed: 3.7
  max_speed: 4.7
  min_speed: 2.6
- time: "2022-10-29T21:52:00+10:00"
  direction: 318
  avg_speed: 3.6
  max_speed: 4.2
  min_speed: 3
- time: "2022-10-29T21:51:00+10:00"
  direction: 317
  avg_speed: 3.9
  max_speed: 5.2
  min_speed: 2.3
- time: "2022-10-29T21:50:00+10:00"
  direction: 327
  avg_speed: 4.5
  max_speed: 5.3
  min_speed: 3.4
- time: "2022-10-29T21:49:00+10:00"
  direction: 316
  avg_speed: 4
  max_speed: 5.1
  min_speed: 3.1
- time: "2022-10-29T21:48:00+10:00"
  direction: 311
  avg_speed: 4.5
  max_speed: 5.6
  min_speed: 3.6
- time: "2022-10-29T21:47:00+10:00"
  direction: 317
  avg_speed: 3.8
  max_speed: 5.1
  min_speed: 3
- time: "2022-10-29T21:46:00+10:00"
  direction: 308
  avg_speed: 3.8
  max_speed: 5.1
  min_speed: 2.8
- time: "2022-10-29T21:45:00+10:00"
  direction: 311
  avg_speed: 4.3
  max_speed: 5.6
  min_speed: 2.1
- time: "2022-10-29T21:44:00+10:00"
  direction: 314
  avg_speed: 4.4
  max_speed: 6
  min_speed: 3.1
- time: "2022-10-29T21:43:00+10:00"
  direction: 308
  avg_speed: 4.5
  max_speed: 5.5
  min_speed: 3.4
- time: "2022-10-29T21:42:00+10:00"
  direction: 320
  avg_speed: 3.5
  max_speed: 5.2
  min_speed: 2.1
- time: "2022-10-29T21:41:00+10:00"
  direction: 307
  avg_speed: 3.1
  max_speed: 4.5
  min_speed: 1.9
- time: "2022-10-29T21:40:00+10:00"
  direction: 325
  avg_speed: 2.9
  max_speed: 4.4
  min_speed: 1.2
- time: "2022-10-29T21:39:00+10:00"
  direction: 289
  avg_speed: 2.8
  max_speed: 4.4
  min_speed: 1.5
- time: "2022-10-29T21:38:00+10:00"
  direction: 282
  avg_speed: 2.6
  max_speed: 3.3
  min_speed: 1.6
- time: "2022-10-29T21:36:00+10:00"
  direction: 316
  avg_speed: 2.7
  max_speed: 4.4
  min_speed: 0.9
- time: "2022-10-29T21:35:00+10:00"
  direction: 334
  avg_speed: 2.4
  max_speed: 3.4
  min_speed: 1.8
- time: "2022-10-29T21:34:00+10:00"
  direction: 315
  avg_speed: 1.6
  max_speed: 2.9
  min_speed: 0.4
- time: "2022-10-29T21:33:00+10:00"
  direction: 325
  avg_speed: 2.4
  max_speed: 3.3
  min_speed: 1.6
- time: "2022-10-29T21:32:00+10:00"
  direction: 307
  avg_speed: 2
  max_speed: 3.3
  min_speed: 1.3
- time: "2022-10-29T21:31:00+10:00"
  direction: 311
  avg_speed: 1.5
  max_speed: 2.3
  min_speed: 0.5
- time: "2022-10-29T21:30:00+10:00"
  direction: 316
  avg_speed: 1.8
  max_speed: 2.6
  min_speed: 1
- time: "2022-10-29T21:29:00+10:00"
  direction: 334
  avg_speed: 2
  max_speed: 2.9
  min_speed: 1.1
- time: "2022-10-29T21:28:00+10:00"
  direction: 312
  avg_speed: 1.9
  max_speed: 2.7
  min_speed: 1.2
- time: "2022-10-29T21:27:00+10:00"
  direction: 302
  avg_speed: 1.3
  max_speed: 1.9
  min_speed: 0.5
- time: "2022-10-29T21:26:00+10:00"
  direction: 321
  avg_speed: 1.3
  max_speed: 2.1
  min_speed: 0.9
- time: "2022-10-29T21:25:00+10:00"
  direction: 321
  avg_speed: 2.2
  max_speed: 2.9
  min_speed: 1.4
- time: "2022-10-29T21:24:00+10:00"
  direction: 313
  avg_speed: 1.6
  max_speed: 2.1
  min_speed: 1
- time: "2022-10-29T21:23:00+10:00"
  direction: 320
  avg_speed: 1.8
  max_speed: 2.5
  min_speed: 1
- time: "2022-10-29T21:21:00+10:00"
  direction: 339
  avg_speed: 1.6
  max_speed: 2.4
  min_speed: 0.9
- time: "2022-10-29T21:20:00+10:00"
  direction: 323
  avg_speed: 1.5
  max_speed: 2.6
  min_speed: 0.7
- time: "2022-10-29T21:19:00+10:00"
  direction: 310
  avg_speed: 1.8
  max_speed: 3.2
  min_speed: 0.7
- time: "2022-10-29T21:18:00+10:00"
  direction: 316
  avg_speed: 2.1
  max_speed: 2.9
  min_speed: 1.2
- time: "2022-10-29T21:17:00+10:00"
  direction: 308
  avg_speed: 1.8
  max_speed: 3.1
  min_speed: 0.8
- time: "2022-10-29T21:16:00+10:00"
  direction: 316
  avg_speed: 1.9
  max_speed: 2.9
  min_speed: 1
- time: "2022-10-29T21:15:00+10:00"
  direction: 318
  avg_speed: 1.6
  max_speed: 2.3
  min_speed: 1
- time: "2022-10-29T21:13:00+10:00"
  direction: 339
  avg_speed: 2.1
  max_speed: 3.1
  min_speed: 1.3
- time: "2022-10-29T21:12:00+10:00"
  direction: 320
  avg_speed: 1.3
  max_speed: 1.9
  min_speed: 0.7
- time: "2022-10-29T21:11:00+10:00"
  direction: 320
  avg_speed: 1.5
  max_speed: 2.3
  min_speed: 0.8
- time: "2022-10-29T21:09:00+10:00"
  direction: 337
  avg_speed: 2.1
  max_speed: 2.9
  min_speed: 1.5
- time: "2022-10-29T21:08:00+10:00"
  direction: 324
  avg_speed: 1.4
  max_speed: 2.9
  min_speed: 0.7
- time: "2022-10-29T21:07:00+10:00"
  direction: 300
  avg_speed: 2.3
  max_speed: 2.8
  min_speed: 1.8
- time: "2022-10-29T21:06:00+10:00"
  direction: 325
  avg_speed: 2.3
  max_speed: 3.6
  min_speed: 1.1
- time: "2022-10-29T21:05:00+10:00"
  direction: 299
  avg_speed: 2
  max_speed: 2.6
  min_speed: 1.2
- time: "2022-10-29T21:03:00+10:00"
  direction: 310
  avg_speed: 1.6
  max_speed: 3.9
  min_speed: 0.8
- time: "2022-10-29T21:01:00+10:00"
  direction: 331
  avg_speed: 1.9
  max_speed: 2.8
  min_speed: 1.2
- time: "2022-10-29T21:00:00+10:00"
  direction: 328
  avg_speed: 2.1
  max_speed: 3
  min_speed: 1.6
- time: "2022-10-29T20:59:00+10:00"
  direction: 313
  avg_speed: 2.2
  max_speed: 2.9
  min_speed: 1.4
- time: "2022-10-29T20:58:00+10:00"
  direction: 335
  avg_speed: 1.3
  max_speed: 2.1
  min_speed: 0.9
- time: "2022-10-29T20:57:00+10:00"
  direction: 321
  avg_speed: 0.9
  max_speed: 1.5
  min_speed: 0.4

